File: VB/FavoriteList/Form1.vb
Imports DevExpress.Utils
Imports DevExpress.XtraBars

Namespace FavoriteList
	Partial Public Class Form1
		Inherits DevExpress.XtraBars.Ribbon.RibbonForm
		Implements IMessageFilter

		Public Sub New()
			InitializeComponent()
			InitToolTipList()
			InitContacts()
			InitTimer()

			Application.AddMessageFilter(Me)
			barManger = ribbonControl1.Manager
		End Sub

		#Region "InitializeComponent"
		Private Sub InitToolTipList()
			ToolTipList.Add(New ToolTipObject(barSubItem3))
			ToolTipList.Add(New ToolTipObject(barSubItem4))
			ToolTipList.Add(New ToolTipObject(barSubItem5))
			ToolTipList.Add(New ToolTipObject(barSubItem6))
		End Sub
		Private Sub InitContacts()
			Dim img As Image = ExampleHelper.GetImage(False)
			barSubItem3.Glyph = img
			barSubItem4.Glyph = img
			barSubItem5.Glyph = img
			barSubItem6.Glyph = img
		End Sub
		Private Sub InitTimer()
			timer = New Timer() With {.Interval = 400}
			AddHandler timer.Tick, AddressOf Tick
		End Sub
		#End Region
		#Region "Fields"
		Private timer As Timer
		Private ReadOnly ToolTipList As New List(Of ToolTipObject)()
		Private ReadOnly barManger As BarManager
		Private ReadOnly itemsAndLinks As New Dictionary(Of BarItem, BarItemLink)()
		Public Const WM_MouseMove As Integer = &H200
		Private isToolTipShown As Boolean = False
		Private link As BarSubItemLink = Nothing
		#End Region
		#Region "Event Handlers"
		Private Sub barSubItem3_ItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs) Handles barSubItem3.ItemClick
			AddToFavorites(e)
		End Sub
		Private Sub barSubItem4_ItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs) Handles barSubItem4.ItemClick
			AddToFavorites(e)
		End Sub

		Private Sub barSubItem5_ItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs) Handles barSubItem5.ItemClick
			AddToFavorites(e)
		End Sub

		Private Sub barSubItem6_ItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs) Handles barSubItem6.ItemClick
			AddToFavorites(e)
		End Sub

		Private Sub Tick(ByVal sender As Object, ByVal e As EventArgs)
			ShowToolTip()
		End Sub
		#End Region

		Private Sub ShowToolTip()
			Dim index As Integer = -1
			Dim item As BarSubItem = Nothing
			For i As Integer = 0 To ToolTipList.Count - 1
				item = ToolTipList(i).ToolTipItem
				link = TryCast(item.Links(0), BarSubItemLink)
				If ExampleHelper.GlyphContainsCursor(link) Then
					index = i
					Exit For
				End If
			Next i

			If index <> -1 Then
				Dim args As New ToolTipControllerShowEventArgs() With {
					.ToolTipLocation = ToolTipLocation.Fixed,
					.SuperTip = New SuperToolTip()
				}
				args.SuperTip.Items.Add(If(itemsAndLinks.ContainsKey(item), ToolTipList(index).RemoveToolTip, ToolTipList(index).AddToolTip))
				Dim linkPoint As Point = ExampleHelper.GetToolTipLocation(link)
				barManger.GetToolTipController().ShowHint(args, linkPoint)
				isToolTipShown = True
			End If
		End Sub

		Private Sub HideToolTip()
			barManger.GetToolTipController().HideHint()
			isToolTipShown = False
		End Sub

		Private Sub HideToolTipIfRequired()
			If isToolTipShown AndAlso Not ExampleHelper.GlyphContainsCursor(link) Then
				HideToolTip()
			End If
		End Sub

		Public Function PreFilterMessage(ByRef m As Message) As Boolean Implements IMessageFilter.PreFilterMessage
			If m.Msg = WM_MouseMove Then
				HideToolTipIfRequired()
				timer.Stop()
				timer.Start()
			End If
			Return False
		End Function

		Private Sub AddToFavorites(ByVal e As ItemClickEventArgs)
			If ExampleHelper.GlyphContainsCursor(TryCast(e.Link, BarSubItemLink)) Then
				Dim inFavorites As Boolean = True
				If barSubItem2.ContainsItem(e.Item) Then
					inFavorites = False
					barSubItem2.ItemLinks.Remove(itemsAndLinks(e.Item))
					itemsAndLinks.Remove(e.Item)
				Else
					itemsAndLinks(e.Item) = barSubItem2.ItemLinks.Add(e.Item)
				End If
				e.Item.Glyph = ExampleHelper.GetImage(inFavorites)
			End If
		End Sub

	End Class
End Namespace

File: CS/FavoriteList/Form1.cs
using DevExpress.Utils;
using DevExpress.XtraBars;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Windows.Forms;

namespace FavoriteList
{
    public partial class Form1 : DevExpress.XtraBars.Ribbon.RibbonForm, IMessageFilter
    {
        public Form1()
        {
            InitializeComponent();
            InitToolTipList();
            InitContacts();
            InitTimer();

            Application.AddMessageFilter(this);
            barManger = ribbonControl1.Manager;
        }

        #region InitializeComponent
        private void InitToolTipList()
        {
            ToolTipList.Add(new ToolTipObject(barSubItem3));
            ToolTipList.Add(new ToolTipObject(barSubItem4));
            ToolTipList.Add(new ToolTipObject(barSubItem5));
            ToolTipList.Add(new ToolTipObject(barSubItem6));
        }
        private void InitContacts()
        {
            Image img = ExampleHelper.GetImage(false);
            barSubItem3.Glyph = img;
            barSubItem4.Glyph = img;
            barSubItem5.Glyph = img;
            barSubItem6.Glyph = img;
        }
        private void InitTimer()
        {
            timer = new Timer() { Interval = 400 };
            timer.Tick += Tick;
        }
        #endregion
        #region Fields
        Timer timer;
        readonly List<ToolTipObject> ToolTipList = new List<ToolTipObject>();
        readonly BarManager barManger;
        readonly Dictionary<BarItem, BarItemLink> itemsAndLinks = new Dictionary<BarItem, BarItemLink>();
        public const int WM_MouseMove = 0x0200;
        bool isToolTipShown = false;
        BarSubItemLink link = null;
        #endregion
        #region Event Handlers
        private void barSubItem3_ItemClick(object sender, ItemClickEventArgs e)
        {
            AddToFavorites(e);
        }
        private void barSubItem4_ItemClick(object sender, ItemClickEventArgs e)
        {
            AddToFavorites(e);
        }

        private void barSubItem5_ItemClick(object sender, ItemClickEventArgs e)
        {
            AddToFavorites(e);
        }

        private void barSubItem6_ItemClick(object sender, ItemClickEventArgs e)
        {
            AddToFavorites(e);
        }

        void Tick(object sender, EventArgs e)
        {
            ShowToolTip();
        }
        #endregion

        private void ShowToolTip()
        {
            int index = -1;
            BarSubItem item = null;
            for (int i = 0; i < ToolTipList.Count; i++)
            {
                item = ToolTipList[i].ToolTipItem;
                link = item.Links[0] as BarSubItemLink;
                if (ExampleHelper.GlyphContainsCursor(link))
                {
                    index = i;
                    break;
                }
            }

            if (index != -1)
            {
                ToolTipControllerShowEventArgs args = new ToolTipControllerShowEventArgs() { ToolTipLocation = ToolTipLocation.Fixed, SuperTip = new SuperToolTip() };
                args.SuperTip.Items.Add(itemsAndLinks.ContainsKey(item) ? ToolTipList[index].RemoveToolTip : ToolTipList[index].AddToolTip);
                Point linkPoint = ExampleHelper.GetToolTipLocation(link);
                barManger.GetToolTipController().ShowHint(args, linkPoint);
                isToolTipShown = true;
            }
        }

        private void HideToolTip()
        {
            barManger.GetToolTipController().HideHint();
            isToolTipShown = false;
        }

        private void HideToolTipIfRequired()
        {
            if (isToolTipShown && !ExampleHelper.GlyphContainsCursor(link))
                HideToolTip();
        }

        public bool PreFilterMessage(ref Message m)
        {
            if (m.Msg == WM_MouseMove)
            {
                HideToolTipIfRequired();
                timer.Stop();
                timer.Start();
            }
            return false;
        }

        private void AddToFavorites(ItemClickEventArgs e)
        {
            if (ExampleHelper.GlyphContainsCursor(e.Link as BarSubItemLink))
            {
                bool inFavorites = true;
                if (barSubItem2.ContainsItem(e.Item))
                {
                    inFavorites = false;
                    barSubItem2.ItemLinks.Remove(itemsAndLinks[e.Item]);
                    itemsAndLinks.Remove(e.Item);
                }
                else
                    itemsAndLinks[e.Item] = barSubItem2.ItemLinks.Add(e.Item);
                e.Item.Glyph = ExampleHelper.GetImage(inFavorites);
            }
        }

    }
}
